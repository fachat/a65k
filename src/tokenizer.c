/****************************************************************************

    tokenizing a line
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

// this code simply tokenizes the input string given some simple basic rules.

#include <ctype.h>
#include <stdbool.h>
#include <string.h>

#include "types.h"
#include "mem.h"
#include "err.h"
#include "tokenizer.h"
#include "log.h"


static type_t tokenizer_memtype = {
	"tokenizer_t",
	sizeof(tokenizer_t)
};

static op_details_t *sorted[256];

static op_details_t ops[] = {
        { 	OP_EXCL,		"!",	1,	0,	0,	0,	0	},
        {	OP_AT,			"@",	1,	0,	0,	0,	0	},
        {	OP_XQUOTE,		"`",	0,	0,	0,	0,	0	},
        {	OP_SQUOTE,		"'",	0,	0,	0,	0,	0	},
        {	OP_OPEN,		"(",	0,	0,	0,	0,	0	},
        {	OP_CLOSE,		")",	0,	0,	0,	0,	0	},
        {	OP_HASH, 		"#",	0,	0,	0,	0,	0	},
        {	OP_BOPEN,		"[",	0,	0,	0,	0,	0	},
        {	OP_BBOPEN,		"[[",	0,	0,	0,	0,	0	},
        {	OP_BCLOSE,		"]",	0,	0,	0,	0,	0	},
        {	OP_BBCLOSE,		"]]",	0,	0,	0,	0,	0	}, 
        {	OP_STAR,		"*",	0,	0,	1,	0,	0	},
        {	OP_COMMA,		",",	0,	0,	0,	0,	0	},
        {	OP_PLUS,		"+",	0,	0,	1,	0,	0	},
        {	OP_MINUS,		"-",	0,	1,	1,	0,	0	},
        {	OP_DIV,			"/",	0,	0,	1,	0,	0	},
        {	OP_COLON,		":",	0,	0,	0,	0,	0	},
        {	OP_SEMICOLON,		";",	0,	0,	0,	0,	0	},
        {	OP_DOT,			".",	0,	0,	0,	0,	0	},
        {	OP_LARGER,		">",	1,	0,	1,	0,	0	},
        {	OP_LARGEROREQUAL, 	">=",	1,	0,	1,	0,	0	},
        {	OP_LESS,		"<",	1,	0,	1,	0,	0	},
        {	OP_LESSOREQUAL,		"<=",	0,	0,	1,	0,	0	},
        {	OP_ASSIGN,		"=",	0,	0,	0,	0,	1	}, 
        {	OP_ASSIGNADD,		"+=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNSUB,		"-=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNMULT,		"*=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNDIV,		"/=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNMOD,		"%=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNBITOR,		"|=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNBITAND,	"&=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNBITXOR,	"^=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNSHIFTLEFT,	"<<=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNSHIFTRIGHT,	">>=",	0,	0,	0,	0,	1	},
        {	OP_BITXOR,		"^",	1,	0,	1,	0,	0	},
        {	OP_BITOR,		"|",	0,	0,	1,	0,	0	},
        {	OP_BITAND,		"&",	0,	0,	1,	0,	0	},
        {	OP_EQUAL,		"==",	0,	0,	1,	0,	0	},
        {	OP_NOTEQUAL,		"!=",	0,	0,	1,	0,	0	},
        {	OP_SHIFTLEFT,		"<<",	0,	0,	1,	0,	0	},
        {	OP_SHIFTRIGHT,		">>",	0,	0,	1,	0,	0	},
        {	OP_LOGICAND,		"&&",	0,	0,	1,	0,	0	},
        {	OP_LOGICOR,		"||",	0,	0,	1,	0,	0	},
        {	OP_LOGICXOR,		"^^",	0,	0,	1,	0,	0	},
        {	OP_BITINV,		"~",	1,	0,	1,	0,	0	},
        {	OP_MOD,			"%",	1,	0,	1,	0,	0	},
        {	OP_DOUBLESLASH,		"//",	0,	0,	0,	0,	0	},
        {	OP_XIND,		",x",	0,	0,	0,	1,	0	},
        {	OP_YIND,		",y",	0,	0,	0,	1,	0	},
        {	OP_ZIND,		",z",	0,	0,	0,	1,	0	},
        {	OP_SIND,		",s",	0,	0,	0,	1,	0	},
        {	OP_BIND,		",b",	0,	0,	0,	1,	0	},
        {	OP_PIND,		",p",	0,	0,	0,	1,	0	},
};

void tokenizer_module_init(void) {

	memset(sorted, 0, 256 * sizeof(op_details_t*));

	unsigned int n = sizeof(ops) / sizeof(op_details_t);

	for (unsigned int i = 0; i < n; i++) {
		op_details_t *p = &ops[i];	

		sorted[p->type] = p;
	}
}


op_details_t *tokenizer_op_details(op_t op) {
	op_details_t *od = sorted[op];

	if (od == NULL) {
		log_fatal("illegal token %d\n", op);
	}
	return od;
}



// initialize a tokenizer 
tokenizer_t *tokenizer_create(const char *line, int cstyle_allowed) {
	
	tokenizer_t *tok = mem_alloc(&tokenizer_memtype);

	tok->line = line;
	tok->cstyle_allowed = cstyle_allowed;
	tok->ptr = 0;
	tok->len = 0;
	tok->type = T_INIT;

	return tok;
}

static bool_t parse_base(tokenizer_t *tok, int ptr, int base) {

	const char *line = tok->line;
	signed char c = 0;

	long value = 0;
	while (line[ptr] != 0) {
		c = tok->line[ptr];
		if (!isalnum(c)) {
			// other tokens like ",", or operators
			break;
		}
		if (isdigit(c)) {
			c &= 0x0f;
		} else {
			c = tolower(c) - 'a' + 10;
		}
		if (c > (base-1)) {
			// TODO: continue till end of number, so simple typo still allows detect further errors
			tok->vals.errno = E_TOK_DIGITRANGE;
			tok->type = T_ERROR;
			return false;
		}
		value = value * base + c;
		ptr++;
	}
	tok->vals.literal.value = value;
	tok->len = ptr - tok->ptr;

	if (tok->len == 0) {
		// syntax error
		tok->vals.errno = E_TOK_EMPTY;
		tok->type = T_ERROR;
		return false;
	}
	return true;
}

static inline bool_t parse_decimal(tokenizer_t *tok, int ptr) {

	tok->type = T_LITERAL;
	tok->vals.literal.type = LIT_DECIMAL;
	return parse_base(tok, ptr, 10);
}

static inline bool_t parse_octal(tokenizer_t *tok, int ptr, int is_c) {

	tok->type = T_LITERAL;
	tok->vals.literal.type = is_c ? LIT_OCTAL_C : LIT_OCTAL;
	return parse_base(tok, ptr, 8);
}

static inline bool_t parse_binary(tokenizer_t *tok, int ptr) {

	tok->type = T_LITERAL;
	tok->vals.literal.type = LIT_BINARY;
	return parse_base(tok, ptr, 2);
}

static inline bool_t parse_hex(tokenizer_t *tok, int ptr, int is_c) {

	tok->type = T_LITERAL;
	tok->vals.literal.type = is_c ? LIT_HEX_C : LIT_HEX;
	return parse_base(tok, ptr, 16);
}

static inline bool_t parse_string(tokenizer_t *tok, int ptr) {

	const char *line = tok->line;

	char delim = line[ptr];
	tok->type = T_STRING;
	tok->vals.string.type = delim;

	ptr++;
	tok->vals.string.ptr = ptr;
	tok->vals.string.len = 0;

	while (isPrint(line[ptr])) {
		if (line[ptr] == delim) {
			break;
		}
		ptr++;
	}
	tok->vals.string.len = ptr - tok->vals.string.ptr;
	
	if (line[ptr] == delim) {
		ptr++;
	} else 
	if (line[ptr] != 0) {
		// non-printable char ends string
		tok->vals.errno = E_TOK_NONPRINT;
		tok->type = T_ERROR;
		return false;
	}

	tok->len = ptr - tok->ptr;

	if (tok->vals.string.len == 0) {
		// empty string
		tok->vals.errno = E_TOK_EMPTY;
		tok->type = T_ERROR;
		return false;
	}
	return true;
}

static inline bool_t parse_name(tokenizer_t *tok, int ptr) {

	const char *line = tok->line;
	tok->type = T_NAME;

	while (isalnum(line[ptr]) || (line[ptr] == '_')) {
		ptr++;
	}

	tok->len = ptr - tok->ptr;

	return true;
}

static inline bool_t parse_token(tokenizer_t *tok, int ptr, int can_have_operator, int allow_index) {

	const char *line = tok->line;

	char c = line[ptr];

	// default outcome
	tok->vals.op = c;
	tok->type = T_TOKEN;
	tok->len = 1;

	switch(c) {
	case '@':
	case '`':
	case '\'':
	case '#':
	case ':':
	case '.':
	case ';':
		return true;
	case '(':
	case ')':
		tok->type = T_BRACKET;
		return true;
	case ',':
		if (allow_index) {
			// check OP_*IND
			char c = tolower(line[ptr+1]);
			if (c == 'x') {
				tok->len++;
				tok->vals.op = OP_XIND;
			} else
			if (c == 'y') {
				tok->len++;
				tok->vals.op = OP_YIND;
			} else
			if (c == 'z') {
				tok->len++;
				tok->vals.op = OP_ZIND;
			} else
			if (c == 'b') {
				tok->len++;
				tok->vals.op = OP_BIND;
			} else
			if (c == 's') {
				tok->len++;
				tok->vals.op = OP_SIND;
			} else
			if (c == 'p') {
				tok->len++;
				tok->vals.op = OP_PIND;
			}
		}
		return true;
	case '%':
		if (line[ptr+1] == '=') {
			tok->vals.op = OP_ASSIGNMOD;
			tok->len++;
			return true;
		}
		return true;
	case '-':
		if (line[ptr+1] == '=') {
			tok->vals.op = OP_ASSIGNSUB;
			tok->len++;
			return true;
		}
		return true;
	case '+':
		if (line[ptr+1] == '=') {
			tok->vals.op = OP_ASSIGNADD;
			tok->len++;
			return true;
		}
		return true;
	case '*':
		if (line[ptr+1] == '=') {
			tok->vals.op = OP_ASSIGNMULT;
			tok->len++;
			return true;
		}
		return true;
	case '/':
		switch (line[ptr+1]) {
		case '=':
			tok->vals.op = OP_ASSIGNDIV;
			tok->len++;
			return true;
		case '/':
			tok->vals.op = OP_DOUBLESLASH;
			tok->len++;
			return true;
		}
		return true;
	case ']':
		tok->type = T_BRACKET;
		if (line[ptr+1] == ']') {
			tok->vals.op = OP_BBCLOSE;
			tok->len++;
			return true;
		}
		// OP_BOPEN
		return true;
	case '[':
		tok->type = T_BRACKET;
		if (line[ptr+1] == '[') {
			tok->vals.op = OP_BBOPEN;
			tok->len++;
			return true;
		}
		// OP_BOPEN
		return true;
	case '!':
		if (can_have_operator) {
			switch (line[ptr+1]) {
			case '=':
				tok->vals.op = OP_NOTEQUAL;
				tok->len++;
				return true;
			default:
				// default
				// OP_EXCL
				return true;
			}
		} else {
			// OP_EXCL
			return true;
		}
	case '>':
		if (can_have_operator) {
			// check ">>", ">=", "><"
			switch (line[ptr+1]) {
			case '>':
				if (line[ptr+2] == '=') {
					tok->vals.op = OP_ASSIGNSHIFTRIGHT;
					tok->len += 2;
					return true;
				}
				tok->vals.op = OP_SHIFTRIGHT;
				tok->len++;
				return true;
			case '=':
				tok->vals.op = OP_LARGEROREQUAL;
				tok->len++;
				return true;
			case '<':
				tok->vals.op = OP_NOTEQUAL;
				tok->len++;
				return true;
			default:
				// default
				return true;
			}
		} else {
			// ">" for high byte modifier or larger than
			return true;
		}
	case '<':
		if (can_have_operator) {
			// check "<<", "<=", "<>"
			switch (line[ptr+1]) {
			case '<':
				if (line[ptr+2] == '=') {
					tok->vals.op = OP_ASSIGNSHIFTLEFT;
					tok->len += 2;
					return true;
				}
				tok->vals.op = OP_SHIFTLEFT;
				tok->len++;
				return true;
			case '=':
				tok->vals.op = OP_LESSOREQUAL;
				tok->len++;
				return true;
			case '>':
				tok->vals.op = OP_NOTEQUAL;
				tok->len++;
				return true;
			default:
				// default
				return true;
			}
		} else {
			// "<" for high byte modifier
			return true;
		}
	case '=':
		if (can_have_operator) {
			// check "==", "=>", "=<"
			switch (line[ptr+1]) {
			case '=':
				tok->vals.op = OP_EQUAL;
				tok->len++;
				return true;
			default:
				// default, assign
				return true;
			}
		} else {
			// '=' assign
			return true;
		}
	case '&':
		// check "&", "&&", "&="
		switch (line[ptr+1]) {
		case '&':
			tok->vals.op = OP_LOGICAND;
			tok->len++;
			return true;
		case '=':
			tok->vals.op = OP_ASSIGNBITAND;
			tok->len++;
			return true;
		} 
		return true;
	case '|':
		// check "|", "||", "|="
		switch (line[ptr+1]) {
		case '|':
			tok->vals.op = OP_LOGICOR;
			tok->len++;
			return true;
		case '=':
			tok->vals.op = OP_ASSIGNBITOR;
			tok->len++;
			return true;
		} 
		return true;
	case '^':
		// check "^", "^^", "^="
		switch (line[ptr+1]) {
		case '^':
			tok->vals.op = OP_LOGICXOR;
			tok->len++;
			return true;
		case '=':
			tok->vals.op = OP_ASSIGNBITXOR;
			tok->len++;
			return true;
		} 
		return true;
	default:
		tok->vals.errno = E_TOK_UNKNOWN;
		tok->type = T_ERROR;
		return false;
	}

	return true;
}

// set to next token; return true when there is a valid token
bool_t tokenizer_next(tokenizer_t *tok, int allow_index) {

	// move behind last token
	int ptr = tok->ptr + tok->len;

	const char *line = tok->line;

	// attempt to not have to store an extra operand/operator flag in the tokenizer	
	// or even have to pass it to this function as parameter...
	bool_t can_have_operator = (tok->type == T_LITERAL) 
			|| (tok->type == T_TOKEN && line[ptr] == ')');

	// skip whitespace
	while (line[ptr] != 0 && isspace(line[ptr])) {
		ptr++;
	}

	tok->ptr = ptr;

	//printf("ptr=%d, c=%d\n", ptr, line[ptr]);

	// are we done with the line yet?
	if (line[ptr] == 0 || tok->type == T_END || tok->type == T_ERROR) {
		tok->len = 0;
		if (tok->type != T_ERROR) {
			tok->type = T_END;
		}
		return false;
	}

	char c = line[ptr];

	if (isdigit(c)) {
		// handle C-style octal (starting with '0'), hex (starting with '0x'), dec
		if (tok->cstyle_allowed && c == '0') {
			if (line[ptr+1] == 'x' || line[ptr+1] == 'X') {
				// hex, C-style
				tok->ptr = ptr;
				ptr += 2;
				return parse_hex(tok, ptr, 1);
			} else {
				// octal
				tok->ptr = ptr;
				ptr += 1;
				return parse_octal(tok, ptr, 1);
			}
		} else {
			// dec
			return parse_decimal(tok, ptr);
		}
	} else
	if ((!can_have_operator) && (c == '%')) {
		// binary
		tok->ptr = ptr;
		ptr++;
		return parse_binary(tok, ptr);
	} else
	if ((!can_have_operator) && (c == '$')) {
		// hex
		tok->ptr = ptr;
		ptr++;
		return parse_hex(tok, ptr, 0);
	} else
	if ((!can_have_operator) && (c == '&')) {
		// octal
		ptr++;
		tok->ptr = ptr;
		return parse_octal(tok, ptr, 0);
	} else
	if (is_string_delim(c)) {
		// string literal
		// c must be of one of quotetype_t enum!
		return parse_string(tok, ptr);
	} else
	if (isalpha(c) || c == '_') {
		// name
		return parse_name(tok, ptr);
	} else
	if (ispunct(c)) {
		// any other token
		return parse_token(tok, ptr, can_have_operator, allow_index);
	} else {
		// non-printable - error
		tok->vals.errno = E_TOK_UNKNOWN;
		tok->type = T_ERROR;
		return false;
	}
	
}

// set to next token; return true when there is a valid name token, that is used as prefix
bool_t tokenizer_next_prefix(tokenizer_t *tok) {

	// move behind last token
	int ptr = tok->ptr + tok->len;

	const char *line = tok->line;

	// skip whitespace
	while (line[ptr] != 0 && isspace(line[ptr])) {
		ptr++;
	}

	tok->ptr = ptr;

	// are we done with the line yet?
	if (line[ptr] == 0 || tok->type == T_END || tok->type == T_ERROR) {
		tok->len = 0;
		if (tok->type != T_ERROR) {
			tok->type = T_END;
		}
		return false;
	}

	return parse_name(tok, ptr);
}

// set to next token; return true when there is a valid name token, that is used as prefix
bool_t tokenizer_next_comment(tokenizer_t *tok, int allow_colon) {

	// move behind last token
	int ptr = tok->ptr + tok->len;

	const char *line = tok->line;

	// skip whitespace
	while (line[ptr] != 0 && isspace(line[ptr])) {
		ptr++;
	}

	tok->ptr = ptr;
	tok->len = 0;

	// are we done with the line yet?
	if (line[ptr] == 0 || tok->type == T_END || tok->type == T_ERROR) {
		if (tok->type != T_ERROR) {
			tok->type = T_END;
		}
		return false;
	}

	tok->len = strlen(tok->line + ptr);
	if (allow_colon) {
		// TODO need escape handling?
		for (int i = 0; i < tok->len; i++) {
			if (line[tok->ptr + i] == ':') {
				tok->len = i;
			}
		}
	}

	return true;
}


void tokenizer_free(tokenizer_t *tok) {
	
	tok->line = NULL;

	mem_free(tok);
}



