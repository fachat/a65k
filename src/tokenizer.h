/****************************************************************************

    tokenizing a line
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef TOKENIZER_H
#define TOKENIZER_H

#include <stdio.h>

#include "types.h"

// token type
typedef enum {
	T_INIT		= 'i',	// initial state
	T_BRACKET	= 'b',	// any type of bracket
	T_NAME		= 'n',	// a name, like label1 do_this
	T_TOKEN		= 't',	// a token, like ( ) . ++ == <> & 
	T_STRING	= 's',	// string
	T_LITERAL	= 'l',	// a numerical literal value (decimal, hex, ...)
	T_ERROR		= 'e',	// an error has happened
	T_END		= 'x',	// end of string reached
} tok_t;

// literal type
typedef enum {
	LIT_NONE		=' ',
	LIT_DECIMAL		='d',	// a decimal number, like 1 2 3
	LIT_OCTAL		='o',	// an octal number, like &127 
	LIT_BINARY		='b',	// a binary number, like %110101
	LIT_HEX			='h',	// a hex number, like $12dA 
	LIT_OCTAL_C		='O',	// C-type an octal number, like 0127
	LIT_HEX_C		='H',	// C-type hex number, like 0x12Ad
	LIT_CHAR		='c',	// single character
	LIT_TWOCHAR		='s',	// double character string
} littype_t;

// quote type for strings
typedef enum {
	Q_SINGLE		= '\'',
	Q_DOUBLE		= '"'
} quotetype_t;

static inline bool is_string_delim(char c) {
	return (c == Q_SINGLE || c == Q_DOUBLE);
} 


typedef enum {
	OP_NONE			= 0,
	OP_EXCL 		= '!',	
	OP_AT			= '@',
	OP_XQUOTE		= '`',
	OP_SQUOTE		= '\'',
	OP_OPEN			= '(',
	OP_CLOSE		= ')',
	OP_HASH			= '#',
	OP_BOPEN		= '[',
	OP_COPEN		= '{',
	OP_BBOPEN		= 136,	// "[["
	OP_BCLOSE		= ']',
	OP_CCLOSE		= '}',
	OP_BBCLOSE		= 137,	// "]]"
	OP_STAR			= '*',
	OP_COMMA		= ',',
	OP_PLUS			= '+',	// binary as usual, as unary modified label scope
	OP_MINUS		= '-',
	OP_DIV			= '/',
	OP_COLON		= ':',
	OP_SEMICOLON		= ';',
	OP_DOT			= '.',
	OP_LARGER		= '>',
	OP_LARGEROREQUAL	= 128,	// ">="
	OP_LESS			= '<',
	OP_LESSOREQUAL		= 129,	// "<="
	OP_ASSIGN		= '=',
	OP_ASSIGNADD		= 138,	// "+="
	OP_ASSIGNSUB		= 139,	// "-="
	OP_ASSIGNMULT		= 140,	// "*="
	OP_ASSIGNDIV		= 141,	// "/="
	OP_ASSIGNMOD		= 142,	// "%="
	OP_ASSIGNBITOR		= 143,	// "|="
	OP_ASSIGNBITAND		= 144,	// "&="
	OP_ASSIGNBITXOR		= 145,	// "^="
	OP_ASSIGNSHIFTLEFT	= 146,	// "<<="
	OP_ASSIGNSHIFTRIGHT	= 147,	// ">>="
	OP_BITXOR		= '^',
	OP_BITOR		= '|',
	OP_BITAND		= '&',	// binary as usual, as unary modifies label scope
	OP_EQUAL		= 138,	// "=="
	OP_NOTEQUAL		= 130,	// "!=", "<>", "><"
	OP_SHIFTLEFT		= 131,	// "<<"
	OP_SHIFTRIGHT		= 132,	// ">>"
	OP_LOGICAND		= 133,	// "&&"
	OP_LOGICOR		= 134,	// "||"
	OP_LOGICXOR		= 135,	// "^^"
	OP_BITINV		= '~',
	OP_MOD			= '%',
	OP_DOUBLESLASH		= 153,	// "//"
	OP_RROPEN		= 154,	// "(("	- not arithm. parsed but generated by admode parser
	OP_RRCLOSE		= 155,	// "))"	- not arithm. parsed but generated by admode parser
	// register indexing
	OP_XIND			= 148, 	// ",x"
	OP_YIND			= 149,	// ",y"
	OP_ZIND			= 150,	// ",z"
	OP_SIND			= 151,	// ",s" - stack offset
	OP_BIND			= 152, 	// ",b" - base register offset
	OP_PIND			= 153, 	// ",p" - PC offset
} op_t;

typedef struct {
	op_t		type;
	const char	*print;
	char		is_modifier;
	char		is_unary;
	char		is_binary;
	char		is_index;	// like ",x"
	char		is_assign;
} op_details_t;


static inline bool is_stmt_delim(op_t t) {
	return (t == OP_COLON);
}

static inline bool is_comment_delim(op_t t) {
	return (t == OP_SEMICOLON) || (t == OP_DOUBLESLASH);
}

static inline op_t closing_op(op_t opening_bracket) {
	switch (opening_bracket) {
	case OP_OPEN: 	return OP_CLOSE;
	case OP_BOPEN:	return OP_BCLOSE;
	case OP_COPEN:	return OP_CCLOSE;
	case OP_BBOPEN:	return OP_BBCLOSE;
	case OP_RROPEN:	return OP_RRCLOSE;
	case OP_LESS:	return OP_LARGER;
	default:	return OP_NONE;
	}
}

extern int prio_of_operator_table[];

static inline int prio_of_operator(const op_t op) {
	return prio_of_operator_table[op];
}

typedef struct {
	const char 	*line;		// line to tokenize
	int		cstyle_allowed;	// config - C-style octal and hex allowed
	tok_t		type;		// type of token
	int 		ptr;		// current pointer to token in line
	int		len;		// length of current token (total, i.e. incl. quotes for a string)
	union {
		struct {
			int		ptr;		// pointer to start of string for T_STRING_LITERAL
			int		len;		// length of string for T_STRING_LITERAL
			quotetype_t	type;		// what quote was used?
		} string;
		struct {
			long		value;		// value for decimal, octal, binary, or hex literal
			littype_t	type;		// type of literal, for T_LITERAL
		} literal;
		op_t op;
		int errno;
	} vals;
} tokenizer_t;

// init the tokenizer
void tokenizer_module_init(void);

// initialize a tokenizer 
tokenizer_t *tokenizer_create(const char *line, int cstyle_allowed);

// set to next token; return true when there is a valid token
// if allow_index is true, comma is folded into ",x" etc addressig mode tokens
// if allow_colon is true, comments may include a colon
bool_t tokenizer_next(tokenizer_t *tok, int allow_index);
bool_t tokenizer_next_prefix(tokenizer_t *tok);
bool_t tokenizer_next_comment(tokenizer_t *tok, int allow_colon);

// clear token, so _next will return the same token again
void tokenizer_rewind(tokenizer_t *tok);

void tokenizer_free(tokenizer_t *tok);

op_details_t *tokenizer_op_details(op_t op);

#endif


