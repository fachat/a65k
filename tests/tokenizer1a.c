
#include <stdbool.h>

#include "log.h"
#include "tokenizer.h"


const char *teststrings[] = {
	"label1 lda #$10",
	"label2: lda <xyz ; comment",
	"dec 125",
	"dec 125,",
	"dec 125;",
	"dec 0128",
	"dec 0127,",
	"dec 0126;",
	"inc 0x12df",
	"inc $1234f",
	"inc 0x1234f,",
	"inc $1234f,",
	"inc $1234f;",
	"inc $12g4f;",
	"inc 0x12g4f;",
	".byt 'xsd&'",
	".byt 'xsd&",
	".asc \"dsdd'",
	".asc \"dsAâ‚¬\"",
	NULL
};


void do_test(const char *teststr);

int main(int argc, char *argv[]) {


	printf("tokenizer:\n");

	for (int p = 0; ;p++) {
		const char *teststr = teststrings[p];
		if (teststr == NULL) {
			break;
		}

		do_test(teststr);
	}
	return 0;
}


void do_test(const char *teststr) {

	printf("tokenizing line: %s\n", teststr);

	tokenizer_t *tok = tokenizer_create(teststr, 0);

	while (tokenizer_next(tok, false)) {

		if (tok->type == T_LITERAL) {
			printf("TOK -> type=%c, len=%d, ptr=%s ltype=%c, val=%ld\n", 
				tok->type, tok->len, tok->line+tok->ptr, tok->vals.literal.type, tok->vals.literal.value);
		} else {
			printf("TOK -> type=%c, len=%d, ptr=%s\n", tok->type, tok->len, tok->line+tok->ptr);
		}
	}
	
	printf("END -> type=%c, rest->%s\n", tok->type, tok->line+tok->ptr);

	tokenizer_free(tok);
}

