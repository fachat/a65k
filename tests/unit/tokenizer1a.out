tokenizer:
tokenizing line: label1 lda #$10
TOK -> type=n, len=6, ptr=label1 lda #$10
TOK -> type=n, len=3, ptr=lda #$10
TOK -> type=t, len=1, ptr=#$10
TOK -> type=l, len=3, ptr=$10 ltype=h, val=16
END -> type=x, rest->
tokenizing line: label2: lda <xyz ; comment
TOK -> type=n, len=6, ptr=label2: lda <xyz ; comment
TOK -> type=t, len=1, ptr=: lda <xyz ; comment
TOK -> type=n, len=3, ptr=lda <xyz ; comment
TOK -> type=t, len=1, ptr=<xyz ; comment
TOK -> type=n, len=3, ptr=xyz ; comment
TOK -> type=t, len=1, ptr=; comment
TOK -> type=n, len=7, ptr=comment
END -> type=x, rest->
tokenizing line: dec 125
TOK -> type=n, len=3, ptr=dec 125
TOK -> type=l, len=3, ptr=125 ltype=d, val=125
END -> type=x, rest->
tokenizing line: dec 125,
TOK -> type=n, len=3, ptr=dec 125,
TOK -> type=l, len=3, ptr=125, ltype=d, val=125
TOK -> type=t, len=1, ptr=,
END -> type=x, rest->
tokenizing line: dec 125;
TOK -> type=n, len=3, ptr=dec 125;
TOK -> type=l, len=3, ptr=125; ltype=d, val=125
TOK -> type=t, len=1, ptr=;
END -> type=x, rest->
tokenizing line: dec 0128
TOK -> type=n, len=3, ptr=dec 0128
TOK -> type=l, len=4, ptr=0128 ltype=d, val=128
END -> type=x, rest->
tokenizing line: dec 0127,
TOK -> type=n, len=3, ptr=dec 0127,
TOK -> type=l, len=4, ptr=0127, ltype=d, val=127
TOK -> type=t, len=1, ptr=,
END -> type=x, rest->
tokenizing line: dec 0126;
TOK -> type=n, len=3, ptr=dec 0126;
TOK -> type=l, len=4, ptr=0126; ltype=d, val=126
TOK -> type=t, len=1, ptr=;
END -> type=x, rest->
tokenizing line: inc 0x12df
TOK -> type=n, len=3, ptr=inc 0x12df
ERR:foo:1 Illegal digit - got 'x', expected base 10
ERR:inc 0x12df
ERR:    ^
END -> type=e, rest->0x12df
tokenizing line: inc $1234f
TOK -> type=n, len=3, ptr=inc $1234f
TOK -> type=l, len=6, ptr=$1234f ltype=h, val=74575
END -> type=x, rest->
tokenizing line: inc 0x1234f,
TOK -> type=n, len=3, ptr=inc 0x1234f,
ERR:foo:1 Illegal digit - got 'x', expected base 10
ERR:inc 0x1234f,
ERR:    ^
END -> type=e, rest->0x1234f,
tokenizing line: inc $1234f,
TOK -> type=n, len=3, ptr=inc $1234f,
TOK -> type=l, len=6, ptr=$1234f, ltype=h, val=74575
TOK -> type=t, len=1, ptr=,
END -> type=x, rest->
tokenizing line: inc $1234f;
TOK -> type=n, len=3, ptr=inc $1234f;
TOK -> type=l, len=6, ptr=$1234f; ltype=h, val=74575
TOK -> type=t, len=1, ptr=;
END -> type=x, rest->
tokenizing line: inc $12g4f;
TOK -> type=n, len=3, ptr=inc $12g4f;
ERR:foo:1 Illegal digit - got 'g', expected base 16
ERR:inc $12g4f;
ERR:    ^
END -> type=e, rest->$12g4f;
tokenizing line: inc 0x12g4f;
TOK -> type=n, len=3, ptr=inc 0x12g4f;
ERR:foo:1 Illegal digit - got 'x', expected base 10
ERR:inc 0x12g4f;
ERR:    ^
END -> type=e, rest->0x12g4f;
tokenizing line: .byt 'xsd&'
TOK -> type=t, len=1, ptr=.byt 'xsd&'
TOK -> type=n, len=3, ptr=byt 'xsd&'
TOK -> type=s, len=6, ptr='xsd&'
END -> type=x, rest->
tokenizing line: .byt 'xsd&
TOK -> type=t, len=1, ptr=.byt 'xsd&
TOK -> type=n, len=3, ptr=byt 'xsd&
TOK -> type=s, len=5, ptr='xsd&
END -> type=x, rest->
tokenizing line: .asc "dsdd'
TOK -> type=t, len=1, ptr=.asc "dsdd'
TOK -> type=n, len=3, ptr=asc "dsdd'
TOK -> type=s, len=6, ptr="dsdd'
END -> type=x, rest->
tokenizing line: .asc "dsA€"
TOK -> type=t, len=1, ptr=.asc "dsA€"
TOK -> type=n, len=3, ptr=asc "dsA€"
TOK -> type=s, len=8, ptr="dsA€"
END -> type=x, rest->
